packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,348 @@
+     * @return UFixed18 wrapped amount for action
+     * @return Current index of calldata to continue decoding
+     */
+    function _decodeAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, UFixed18 amount, uint) {
+        (addr1, ptr) = _decodeAddress(input, ptr);
+        (amount, ptr) = _decodeAmountUFixed18(input, ptr);
+
+        return(addr1, amount, ptr);
+    }
+
+    /**
+     * @notice Decodes next bytes of action as address, uint256[]
+     * @param input Full calldata payload
+     * @param ptr Current index of input to start decoding
+     * @return Product address for action
+     * @return ProgramIds for action
+     * @return Current index of calldata to continue decoding
+     */
+    function _decodeProductPrograms(bytes calldata input, uint ptr) private returns (address product, uint256[] memory programs, uint) {
+        (product, ptr) = _decodeAddress(input, ptr);
+        (programs, ptr) = _decodeUintArray(input, ptr);
+
+        return(product, programs, ptr);
+    }
+
+    /** 
+     * @notice decodes an address from calldata 
+     * @dev if length == 0, stores next 20 bytes as address to cache 
+     * else loading address from uint cache index
+     * @param input Full calldata payload
+     * @param ptr Current index of input to start decoding
+     * @return Address encoded in calldata
+     * @return Current index of calldata to continue decoding
+    */
+    function _decodeAddress(bytes calldata input, uint ptr) private returns (address addr, uint) {
+        uint8 addrLen = _bytesToUint8(input[ptr:ptr+1]);
+        ptr += 1;
+
+        // user is new to registry, add next 20 bytes as address to registry and return address
+        if (addrLen == 0) {
+            addr = _bytesToAddress(input[ptr:ptr+20]);
+            ptr += 20;
+
+            _setAddressCache(addr);
+
+        } else {
+            uint addrNonceLookup = _bytesToUint256(input[ptr:ptr+addrLen]);
+            ptr += addrLen;
+
+            addr = _getAddressCacheSafe(addrNonceLookup);
+        }
+
+        return (addr, ptr);
+    }
+
+    /**
+     * @notice checked gets the address in cache mapped to the cache index
+     * @dev there is an issue with the calldata if a txn uses cache before caching address
+     * @param nonce The cache index
+     * @return Address stored at cache index
+     */
+    function _getAddressCacheSafe(uint nonce) private view returns (address addr){
+        addr = addressCache[nonce];
+        if (addr == address(0)) revert MultiInvokerRollupInvalidCalldataError();
+    }
+
+    /**
+     * @notice wraps next length of bytes as UFixed18
+     * @param input Full calldata payload
+     * @param ptr Current index of input to start decoding
+     * @param ptr Current index of input to start decoding
+     * @return Current index of calldata to continue decoding
+     */
+     function _decodeAmountUFixed18(bytes calldata input, uint ptr) private view returns (UFixed18 result, uint) {
```

Comment:
Idea to clean up this pointer logic:

1. Create a memory struct for the pointer
```
struct PTR {
    uint256 pos;
}
```

2. Instantiate once in top-level `_decodeFallbackAndInvoke` then simply pass this around into the various decode functions.

This will allow you encapsulate all of the pointer incrementing logic inside of the individual decode functions instead of having to return and reset the point throughout the entire contract.